common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/IdentityState.scala

```diff
@@ -1,41 +1,50 @@
 package uk.ac.wellcome.models.work.internal
 
-sealed trait IdentityState[+T]
-
-sealed trait Unminted[+T] extends IdentityState[T] {
-  val agent: T
-  def withAgent[A](f: T => A): Unminted[A]
+/** The root trait for all possible ID types. */
+sealed trait IdState {
+  def maybeCanonicalId: Option[String]
+  def allSourceIdentifiers: List[SourceIdentifier]
 }
 
-sealed trait Minted[+T] extends IdentityState[T] {
-  val agent: T
-  def withAgent[A](f: T => A): Minted[A]
+/** Parent trait for an ID of an object that is pre minter. */
+sealed trait Unminted extends IdState
+
+/** Parent trait for an ID of an object that is post minter. */
```

Comment:
My only confusing, which predates this PR, is that `Minted` is the state that it has passed through the minter, rather than being minted. PreMinter / PostMinter, as the comment says might make more sense, but, bah.

Reply:
I still don't see though where this is useful?

Reply:
It is useful as anywhere we have an object which is parameterised by `Minted` (or vica versa with `Unminted`), we have made a guarantee that the object has either had a `canonicalId` assigned, or is an object that is unidentifiable. This helps from both the standpoint about reasoning about our code and ensuring it is correct, and reducing the verbosity (for example not having to pattern match against a not yet identified object in such cases, which is something which should never occur within certain parts of the application)